fix Literally undefined class #3283

bre30kra69cs opened the pull request:
**Literally undefined class #3283**

Added check for className. In compiled file its look like:

![image](https://user-images.githubusercontent.com/12611926/62211023-319fa580-b3a6-11e9-81a2-950951786a45.png)

Its no option to modify ‘attr’ function because it gets concatenated className string from image above (null + " svelte-abc"):

![image](https://user-images.githubusercontent.com/12611926/62211128-6d3a6f80-b3a6-11e9-9d85-bae30a53fbd8.png):

I prefer logical 'or' to ternary operator because className token may be a function with side effects and we will call function twice:

![image](https://user-images.githubusercontent.com/12611926/62211344-d9b56e80-b3a6-11e9-9f00-254676cd8101.png)
![image](https://user-images.githubusercontent.com/12611926/62211353-e0dc7c80-b3a6-11e9-8429-84bf97a6fe65.png)

I think in future we must add special class name resolver function. Also added tests.
Thanks

# update
What we now have:
- simplified class name calc logic in compile phase;
- added class name resolver in runtime;
- resolved all problems with falsy values;
- added tests "func token" and no-var;
- added logic where class name resolver will be used only in situations with styled component and one addition token to class hash;

cvlab commented:
@bre30kra69cs It should filter only `undefined` and `null` values, but not `false` or `0`

```JS
${renderedChunk} == null ? '' : ${renderedChunk}
```

cvlab commented:
@bre30kra69cs I guess that 

```html
<div class={testName1 + testName2}></div>
```

is not a very good example because it will do `+` **before** checking if it's `undefined` or `null`, should be testing an variable without operation `+`

```html
<div class={testName1}></div>
```

or 

```html
<div class={testName} class:MyClass={ testVar }></div>
```

bre30kra69cs commented:
Got it!

bre30kra69cs commented:
What we now have:
- simplified class name calc logic in compile phase;
- added class name resolver in runtime;
- resolved all problems with falsy values;
- added tests "func token" and no-var;
- added logic where class name resolver will be used only in situations with styled component and one addition token to class hash;

bre30kra69cs commented:
What we now have:
- simplified class name calc logic in compile phase;
- added class name resolver in runtime;
- resolved all problems with falsy values;
- added tests "func token" and no-var;
- added logic where class name resolver will be used only in situations with styled component and one addition token to class hash;

Rich-Harris commented:
thank you! i made some additional changes in #3330, so that it works for SSR as well
